src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -47,9 +47,42 @@ public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> so
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.size);
             }
 
+            if (source.TryGetNonEnumeratedCount(out int count))
+            {
+                return ChunkIterator(source, size, count);
+            }
```

Comment:
> Was inspired to work on this after seeing https://github.com/dotnet/runtime/pull/72811

The design of this method, e.g. returning a new array for every chunk, is inherently not high-performance, nor is there evidence that improvements to the performance of this method make a material difference in any application, nor data that it's frequently used with countable sources.  Given all that, we've purposefully avoided spending any additional code on trying to optimize various cases, e.g. duplicating the amount of code required to special-case countable sources.  The purpose of #72811 was purely to help mitigate a regression from a previous fix in .NET 7, which is different from trying to optimize the method beyond what shipped initially.  I appreciate your efforts on the PR, but unless there's new data suggesting this is really meaningful, I don't believe it's worth the tradeoff in complexity and manageability.